cjworkbench/i18n/trans.py

```diff
@@ -192,31 +200,53 @@ def for_module_zipfile(cls, module_zipfile: ModuleZipfile):
 
 class MessageLocalizerRegistry:
     def __init__(self):
-        self._supported_modules = WeakValueDictionary()
         self._module_localizers = WeakKeyDictionary()
         self._module_localizers_lock = threading.Lock()
         self._app_localizer = MessageLocalizer.for_application()
 
-    def for_module_id(self, module_id: str) -> MessageLocalizer:
-        with self._module_localizers_lock:
-            return self._module_localizers[self._supported_modules[module_id]]
+    def for_module_zipfile(self, module_zipfile: ModuleZipfile) -> MessageLocalizer:
+        """Return a `MessageLocalizer` for the given `ModuleZipFile`
 
-    def for_application(self):
-        return self._app_localizer
+        Raise `NotInternationalizedError` if the `ModuleZipFile` has no po files
+        
+        Caches the result for each `ModuleZipFile`.
+        """
+        # 1. Get without locking.
+        try:
+            result = self._module_localizers[module_zipfile]
+            if isinstance(result, NotInternationalizedError):
+                raise result
+            else:
+                return result
+        except KeyError:
+            pass  # fall through
 
-    def update_supported_modules(self):
+        # 2. Lock, and try again. (This prevents a race.)
         with self._module_localizers_lock:
-            for module_id, module_zipfile in MODULE_REGISTRY.all_latest().items():
-                if module_zipfile not in self._module_localizers:
-                    self._supported_modules[module_id] = module_zipfile
-                    self._module_localizers[
-                        module_zipfile
-                    ] = MessageLocalizer.for_module_zipfile(module_zipfile)
-
-    def clear(self):
-        self._supported_modules.clear()
-        self._module_localizers.clear()
-        self._app_localizer = MessageLocalizer.for_application()
+            try:
+                result = self._module_localizers[module_zipfile]
+                if isinstance(result, NotInternationalizedError):
+                    raise result
+                else:
+                    return result
+            except KeyError:
+                pass  # we'll fetch it now
+
+            # 3. Update the cache, still holding the lock.
+            try:
+                localizer = MessageLocalizer.for_module_zipfile(module_zipfile)
```

Comment:
Please nix `MessageCatalogRegistry.for_module_zipfile()` and `MessageCatalog.for_module_zipfile()` and ... well ... _any_ other `.for_module_zipfile()` ... and use a private method on this class.

`MessageLocalizerRegistry` is a factory. It should not use factory methods from other classes.